routes/profiles.rb

```diff
@@ -50,13 +51,29 @@ def self.registered(app)
           end
 
           update_profile = lambda do
-            if !@profile || @profile[:role] != 'superadmin'
-              halt 400, 'Parameter requirements were not met.' unless check_parameters(params, profile_params)
+            updatedProfile = params[:updatedProfile]
+            if !@profile || @profile[:id].to_s != updatedProfile[:_id][:$oid].to_s
+              if @profile[:role] != 'superadmin'
+                halt 400, 'Permission Denied.' unless check_parameters(params, profile_params)
+              end
+            end
+            if !check_parameters(updatedProfile, profile_params)
+              halt 400, 'Parameter requirements were not met.'
+            else
+              if updatedProfile[:change_password] == true
+                if !check_password(@profile[:password_hash], updatedProfile[:oldPassword])
+                  halt 400, 'Invalid Credentials. Permission Denied.'
```

Comment:
A couple of things:
- This should be a 401.
- We need to combine all of the auth checks in one area and give superadmins a free pass through it. This is the third implicit or explicit auth check.

Reply:
The intention of this route is to be used on the update profile section - when a user attempts to edit their own profile. It is setup to optionally change a password, but a password change is not required. So I put this authorization separately.

But I think I see what you mean by moving the authorizations together. Let me see if I can come up with something that will handle these situations.

Reply:
I combined the if logic in what I hope is a more efficient manner, and placed it in the helper function - but I had to keep the superadmin check with the password since there is still the possibility a user may update other parts of the profile without attempting to change their password